### PR TITLE
NeuralMI v2.1.0 — variational wrapper, IB decoder, rigorous CI/PI, pairwise heatmap

### PR DESCRIPTION
**Pull request by eslam-abdelaleem**
## Summary

Pre-deployment audit of `feature/variational-wrapper` → NeuralMI v2.1.0.
Bug fixes, new features, documentation, and a full test suite (282 passed, 1 skipped).

## Bug Fixes

- **KL double-normalization** (`critics.py`): `VariationalWrapper` already returns per-sample KL; chunked-eval path was dividing by `batch_size` a second time
- **`infonce` dtype error** (`bounds.py`): `float()` cast prevents `int64` being passed to `torch.log`
- **Decoder categorical loss** (`trainer.py`): `softmax` activation now uses NLL loss; `linear`/`sigmoid` use MSE

## New Features

- **Dual CI / PI** (`rigorous.py`): `_extrapolate_mi` returns `(intercept, mi_error, mi_error_pred, slope)` — both CI and PI half-widths surfaced in `Results.summary()`
- **Negative-estimate warnings** (`conditional.py`, `transfer.py`): `UserWarning` when CMI or TE is negative
- **Embedding-dimension ceiling warning** (`dimensionality.py`): `UserWarning` when participation ratio ≥ 80 % of `embedding_dim`
- **Pairwise heatmap** (`results.py`): `Results.plot(mode='pairwise')` renders a seaborn heatmap with diagonal masking
- **Decoder activation routing** (`trainer.py`, `task.py`): `decoder_output_activation_x/y` wired end-to-end from `base_params` to `Trainer._decoder_loss()`

## Config / Schema

- `pyproject.toml`: added `[build-system]` table (PEP 517); `scikit-learn` → optional `[sklearn]`/`dev` extras
- `defaults.py`: `pairwise` in `MODE_KWARGS_SCHEMA`; `corrupt_target='both'` comment
- `visualize/plot.py`: helpful `ImportError` → `pip install neural_mi[sklearn]`

## Documentation

- `THEORY.md`: §8 Information Bottleneck Extension; §9 `train_mi` as primary MI estimate; β normalisation note
- `run.py`: `history_window`, `prediction_horizon`, `bidirectional_te`, `corrupt_target='both'`, `base_params` vs `**analysis_kwargs`

## Tests

- `test_trainer.py`: `TestDecoderLoss` (4 cases)
- `test_precision.py`: `corrupt_target='both'` end-to-end
- `test_results_extended.py`: `mi_error_pred`; pairwise heatmap
- `test_workflow_internals.py`: 4-tuple unpack for `_extrapolate_mi`

## Test plan

- [ ] `pytest` → 282 passed, 1 skipped, 0 failures
- [ ] `pip install -e .` on clean env (no sklearn import error)
- [ ] `pip install -e ".[sklearn]"` adds scikit-learn
- [ ] `Results.plot(mode='pairwise')` renders on toy `mi_matrix`
- [ ] Dimensionality ceiling warning fires when `embedding_dim` is set low

### CODE REVIEW

File: neural_mi/models/critics.py [hunk 1]
```diff
@@ -39,47 +39,52 @@ def _get_embeddings_and_kl(self, x_out: any, y_out: any) -> Tuple[torch.Tensor,
...
-        
-        if use_variational and not isinstance(total_kl_raw, float):
-            # Single normalization by full batch size — consistent with fast path
-            total_kl = total_kl_raw.to(x_embedded.device) / batch_size
+
+        if use_variational and n_chunks > 0:
+            # Average per-sample-mean KL across chunks for a consistent per-sample estimate.
+            total_kl = total_kl_acc.to(x_embedded.device) / n_chunks
         else:
             total_kl = torch.tensor(0.0, device=x_embedded.device)
```
[the remark this answers is not in the document]
Comment (eslam-abdelaleem) [reply neural_mi/models/critics.py:90]: @copilot apply changes based on [this feedback](https://github.com/eslam-abdelaleem/NeuralMI/pull/16#discussion_r3017227919)

File: neural_mi/models/critics.py [hunk 2]
```diff
@@ -191,31 +242,35 @@ def __init__(self,
...
-                
+
             scores[i:end_idx] = chunk_scores.squeeze()
 
-        if self.use_variational and not isinstance(total_kl_raw, float):
-            kl_tensor = total_kl_raw.to(x.device) / (batch_size * batch_size)
+        if self.use_variational and n_pair_chunks > 0:
+            # Average per-chunk per-sample-mean KL across pair chunks.
+            kl_tensor = total_kl_acc.to(x.device) / n_pair_chunks
         else:
```
[the remark this answers is not in the document]
Comment (eslam-abdelaleem) [reply neural_mi/models/critics.py:281]: @copilot apply changes based on [this feedback](https://github.com/eslam-abdelaleem/NeuralMI/pull/16#discussion_r3017227939)

File: neural_mi/visualize/plot.py [hunk 3]
```diff
@@ -222,10 +222,20 @@ def plot_embeddings(
             )
         z_plot = z[:, :dim]
     elif method == 'pca':
-        from sklearn.decomposition import PCA
+        try:
+            from sklearn.decomposition import PCA
+        except ImportError:
+            raise ImportError(
+                "PCA requires scikit-learn. Install it with: pip install neural_mi[sklearn]"
+            )
```
[the remark this answers is not in the document]
Comment (eslam-abdelaleem) [reply neural_mi/visualize/plot.py:230]: @copilot apply changes based on [this feedback](https://github.com/eslam-abdelaleem/NeuralMI/pull/16#discussion_r3017228048)

File: neural_mi/models/decoders.py [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+        window_size=window_size,
+        n_layers=n_layers,
+        output_activation=output_activation,
+    )
+    name = embedding_model.lower()
+    if name == 'mlp':
+        return MLPDecoder(**common)
+    elif name == 'cnn1d':
+        return CNN1DDecoder(**common, kernel_size=kwargs.get('kernel_size', 7))
+    elif name == 'gru':
```
[the remark this answers is not in the document]
Comment (eslam-abdelaleem) [reply neural_mi/models/decoders.py:425]: @copilot apply changes based on [this feedback](https://github.com/eslam-abdelaleem/NeuralMI/pull/16#discussion_r3017228078)
